server/api/groups.js

```diff
@@ -0,0 +1,250 @@
+    data: { name: group.name },
+    ip: ctx.request.ip,
+  });
+
+  ctx.body = {
+    data: presentGroup(group),
+    policies: presentPolicies(user, [group]),
+  };
+});
+
+router.post('groups.update', auth(), async ctx => {
+  const { id, name } = ctx.body;
+  ctx.assertPresent(name, 'name is required');
+  ctx.assertUuid(id, 'id is required');
+
+  const user = ctx.state.user;
+  const group = await Group.findByPk(id);
+
+  authorize(user, 'update', group);
+
+  group.name = name;
+  await group.save();
+
+  await Event.create({
+    name: 'groups.update',
+    teamId: user.teamId,
+    actorId: user.id,
+    data: { name },
+    ip: ctx.request.ip,
+  });
+
+  ctx.body = {
+    data: presentGroup(group),
+    policies: presentPolicies(user, [group]),
+  };
+});
+
+router.post('groups.delete', auth(), async ctx => {
+  const { id } = ctx.body;
+  ctx.assertUuid(id, 'id is required');
+
+  const { user } = ctx.state;
+  const group = await Group.findByPk(id);
+
+  authorize(user, 'delete', group);
+  await group.destroy();
+
+  await Event.create({
+    name: 'groups.delete',
+    actorId: user.id,
+    teamId: group.teamId,
+    data: { name: group.name },
+    ip: ctx.request.ip,
+  });
+
+  ctx.body = {
+    success: true,
+  };
+});
+
+router.post('groups.memberships', auth(), pagination(), async ctx => {
+  const { id, query, permission } = ctx.body;
+  ctx.assertUuid(id, 'id is required');
+
+  const user = ctx.state.user;
+  const group = await Group.findByPk(id);
+
+  authorize(user, 'read', group);
+
+  let where = {
+    groupId: id,
+  };
+
+  let userWhere;
+
+  if (query) {
+    userWhere = {
+      name: {
+        [Op.iLike]: `%${query}%`,
+      },
+    };
+  }
+
+  if (permission) {
+    where = {
+      ...where,
+      permission,
+    };
+  }
+
+  const memberships = await GroupUser.findAll({
+    where,
+    order: [['createdAt', 'DESC']],
+    offset: ctx.state.pagination.offset,
+    limit: ctx.state.pagination.limit,
+    include: [
+      {
+        model: User,
+        as: 'user',
+        where: userWhere,
+        required: true,
+      },
+    ],
+  });
+
+  ctx.body = {
+    pagination: ctx.state.pagination,
+    data: {
+      groupMemberships: memberships.map(presentGroupMembership),
+      users: memberships.map(membership => presentUser(membership.user)),
+    },
+  };
+});
+
+router.post('groups.add_user', auth(), async ctx => {
+  const { id, userId } = ctx.body;
+  ctx.assertUuid(id, 'id is required');
+  ctx.assertUuid(userId, 'userId is required');
+
+  const group = await Group.findByPk(id);
+  authorize(ctx.state.user, 'update', group);
+
+  const user = await User.findByPk(userId);
+  authorize(ctx.state.user, 'read', user);
+
+  let membership = await GroupUser.findOne({
+    where: {
+      groupId: id,
+      userId,
+    },
+  });
+
+  if (!membership) {
+    membership = await group.addUser(user, {
+      through: { createdById: ctx.state.user.id },
+    });
+  }
+
+  await Event.create({
```

Comment:
This event wants to be moved within the `!membership` block above, right?

Reply:
yup yup